### PR TITLE
WIP add rewrite file operator after iceberg committer

### PR DESCRIPTION
**Pull request by simonsssu**
Purpose: When running Flink Iceberg SInk, we may generate a lot of Datafiles because frequently commit. we have to control the amount of datafiles when we execute a streaming job.
This WIP pr appends rewrite datafile operator after iceberg committer, we can trigger rewrite action when running the flink streaming job. BTW, this can handle a small amount of datafile when a stream job is running.  and we don't worried to trigger a rewrite action manually.  We can also trigger a rewrite action offline.

### CODE REVIEW

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java [hunk 1]
```diff
@@ -179,37 +186,93 @@ public Builder writeParallelism(int newWriteParallelism) {
         try (TableLoader loader = tableLoader) {
           this.table = loader.loadTable();
         } catch (IOException e) {
-          throw new UncheckedIOException("Failed to load iceberg table from table loader: " + tableLoader, e);
+          throw new UncheckedIOException(
+              "Failed to load iceberg table from table loader: " + tableLoader,
+              e);
```
Comment (zhangjun0x01) [inline flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:191]: it seems no change

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java [hunk 2]
```diff
@@ -179,37 +186,93 @@ public Builder writeParallelism(int newWriteParallelism) {
...
-          .transform(ICEBERG_STREAM_WRITER_NAME, TypeInformation.of(DataFile.class), streamWriter)
-          .setParallelism(writeParallelism)
-          .transform(ICEBERG_FILES_COMMITTER_NAME, Types.VOID, filesCommitter)
+      DataStream<?> returnStream = rowDataInput
+          .transform(ICEBERG_STREAM_WRITER_NAME,
+              TypeInformation.of(DataFile.class), streamWriter)
+          .setParallelism(rowDataInput.getParallelism())
+          .transform(ICEBERG_FILES_COMMITTER_NAME,
+              TypeInformation.of(Long.class), filesCommitter)
           .setParallelism(1)
```
Comment (zhangjun0x01) [inline flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:204]: I feel that the master is not the latest version when the patch is rebase

**Comment by zhangjun0x01** [comment]
In the case of a large amount of data, in order to ensure the performance of normal writing,
1. Should we use asynchronous Rewrite? (use AsyncWaitOperator)
2. Should you add a switch to control whether to  enable rewrite

**Comment by rdblue** [comment]
@openinx, please let me know if/when you'd like me to take a look at this PR.

**Comment by openinx** [comment]
Will take a look today, thanks.

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java [hunk 3]
```diff
@@ -179,37 +186,93 @@ public Builder writeParallelism(int newWriteParallelism) {
...
-      this.writeParallelism = writeParallelism == null ? rowDataInput.getParallelism() : writeParallelism;
-
-      DataStream<Void> returnStream = rowDataInput
-          .transform(ICEBERG_STREAM_WRITER_NAME, TypeInformation.of(DataFile.class), streamWriter)
-          .setParallelism(writeParallelism)
-          .transform(ICEBERG_FILES_COMMITTER_NAME, Types.VOID, filesCommitter)
+      DataStream<?> returnStream = rowDataInput
+          .transform(ICEBERG_STREAM_WRITER_NAME,
+              TypeInformation.of(DataFile.class), streamWriter)
+          .setParallelism(rowDataInput.getParallelism())
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:201]: Why change this ? we will need to allow users to set their own writer parallelism, so I think we may need to revert this ?

File: flink/src/main/java/org/apache/iceberg/flink/FlinkTableProperties.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Table Properties when using Flink IcebergSink.
+ */
+public class FlinkTableProperties {
+
+  private FlinkTableProperties() {
+  }
+
+  public static final String CASE_SENSITIVE = "case-sensitive";
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/FlinkTableProperties.java:30]: Those properties are provided for flink only ? we'd better to have a `flink.` prefix in the keys.

File: flink/src/main/java/org/apache/iceberg/flink/sink/rewrite/AbstractRewriteOperator.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Multimaps;
+import org.apache.iceberg.relocated.com.google.common.collect.Streams;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.apache.iceberg.util.TableScanUtil;
+
+abstract class AbstractRewriteOperator<InT, OutT> extends AbstractStreamOperator<OutT>
+    implements OneInputStreamOperator<InT, OutT>, BoundedOneInput {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/sink/rewrite/AbstractRewriteOperator.java:53]: We've abstracted the `BaseRewriteDataFilesAction` in PR https://github.com/apache/iceberg/pull/1624, I think it's good to reuse that part of code.  we could abstract it again when it's necessary.

File: flink/src/main/java/org/apache/iceberg/flink/sink/rewrite/RowDataRewriter.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.flink.sink.TaskWriterFactory;
+import org.apache.iceberg.flink.source.RowDataIterator;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.TaskWriter;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RowDataRewriter implements Serializable {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/sink/rewrite/RowDataRewriter.java:37]: We will have a `Rewriter` once the  https://github.com/apache/iceberg/pull/1623/files#diff-81a2df27390859ae0b203aec87391f01a201ea07d2e7226a9e84e87f74dd3e5fR50  get merged,   so will need to consider sharing code between them.

**Comment by JingsongLi** [comment]
Thanks @simonsssu for your contribution.
If you don't mind, can you write a design document?
I think it is good to have a global design.

I have some initial question:
- When to do compaction?
- Does compaction block checkpoint? If there are large number of files need to be compacted, that will lead to the timeout/failure of checkpoint.
- How to do incremental compaction instead of compacting the whole table?
- When to commit compaction? Does it support multiple streaming sinks?
- in addition, can we reuse more codes in https://github.com/apache/iceberg/pull/1623 ? I think essentially they do the same job.

**Comment by simonsssu** [comment]
@JingsongLi Hi Jingsong, Sorry for late response.  I will write a design doc in google doc.

File: flink/src/main/java/org/apache/iceberg/flink/sink/rewrite/RowDataRewriter.java (hunk 6 above)
Comment (simonsssu) [reply flink/src/main/java/org/apache/iceberg/flink/sink/rewrite/RowDataRewriter.java:37]: Yes, we can reuse most of logic here.

**Comment by simonsssu** [comment]
https://docs.google.com/document/d/1pK2aN9Nv1HjYaJFYTQUWsttp4ahwFPoZbvmPpZ0T32k/edit#heading=h.asry5i6qtp9p
@JingsongLi  Hi Jingsong, I have draft a simple doc. Sorry for late response.

**Comment by openinx** [comment]
@simonsssu Mind to grant the access to the document ?

**Comment by simonsssu** [comment]
@openinx @JingsongLi Done 
https://docs.google.com/document/d/1pK2aN9Nv1HjYaJFYTQUWsttp4ahwFPoZbvmPpZ0T32k/edit?usp=sharing

**Pull request closed without merging**